File: trunk/Geometry/partial_edge.cpp
#ifndef PARTEDGE_H_
#define PARTEDGE_H_

#include <cmath>
#include <map>
#include <new>
using namespace std;

#include "geoquant.h"
#include "triposition.h"

#include "geoquants.h"
#include "triangulation/triangulation.h"

PartialEdgeIndex* PartialEdge::Index = NULL;

PartialEdge::PartialEdge( Vertex& v, Edge& e  ){    
    length = Length::At(e);
    length->addDependent(this);

    radA = Radius::At(v);
    radA->addDependent(this);
     
    
    int v2Index;
    if((*(e.getLocalVertices()))[0] != v.getIndex())
    {
       v2Index = (*(e.getLocalVertices()))[0];
    } else
    {
       v2Index = (*(e.getLocalVertices()))[1];
    }
    
    radB = Radius::At(Triangulation::vertexTable[v2Index]);
    radB->addDependent(this);  
}

void PartialEdge::remove() {
    deleteDependents();
    length->removeDependent(this);
    radA->removeDependent(this);
    radB->removeDependent(this);
    Index->erase(pos);
    delete this;
}

PartialEdge::~PartialEdge(){ 
    
}
void PartialEdge::recalculate() {
   double len = length->getValue();
   value = (pow(len, 2) + pow(radA->getValue(), 2) 
           - pow(radB->getValue(), 2)) / (2 * len);
}

PartialEdge* PartialEdge::At( Vertex& v, Edge& e  ){
  TriPosition T( 2, v.getSerialNumber(), e.getSerialNumber() );
  if( Index == NULL ) Index = new PartialEdgeIndex();
  PartialEdgeIndex::iterator iter = Index->find( T );

  if( iter == Index->end() ){
    PartialEdge* val = new PartialEdge( v, e );
    val->pos = T;
    Index->insert( make_pair( T, val ) );
    return val;
  } else {
    return iter->second;
  }
}

void PartialEdge::CleanUp(){
  if( Index == NULL ) return;
  PartialEdgeIndex::iterator iter;
  PartialEdgeIndex copy = *Index;
  for(iter = copy.begin(); iter != copy.end(); iter++) {
    iter->second->remove();
  }
    
  delete Index;
  Index = NULL;
}

#endif /* PARTEDGE_H_ */

File: trunk/Java/src/geo/PartialEdge.java
package geo;

import java.util.HashMap;

import triangulation.*;


public class PartialEdge extends Geoquant {
  // Index map
  private static HashMap<TriPosition, PartialEdge> Index = new HashMap<TriPosition, PartialEdge>();
  private HashMap<TriPosition, Partial> PartialIndex;
  private HashMap<TriPosition, SecondPartial> SecondPartialIndex;
  // Needed geoquants
  private Radius ri;
  private Radius rj;
  private Alpha ai;
  private Alpha aj;
  private Length Lij;
  private Edge e;
  private Vertex v;
  
  public PartialEdge(Vertex vertex, Edge edge) {
    super(vertex, edge);
    PartialIndex = new HashMap<TriPosition, Partial>();
    SecondPartialIndex = new HashMap<TriPosition, SecondPartial>();
    
    this.e = edge;
    this.v = vertex;
    ri = Radius.At(vertex);
    ai = Alpha.At(vertex);
    for (Vertex v : edge.getLocalVertices()) {
      if (!v.equals(vertex)) {
        rj = Radius.At(v);
        aj = Alpha.At(v);
        break;
      }
    }
    
    Lij = Length.At(edge);
    
    ri.addObserver(this);
    rj.addObserver(this);
    ai.addObserver(this);
    aj.addObserver(this);
    Lij.addObserver(this);
    
  }
  
  protected void recalculate() {
    double length = Lij.getValue();
    double radi = ri.getValue();
    double radj = rj.getValue();
    double alphai = ai.getValue();
    double alphaj = aj.getValue();
    
    value = (length*length + alphai*radi*radi - alphaj*radj*radj) / (2 * length);
  }
 
  public void remove() {
    deleteDependents();
    Lij.deleteObserver(this);
    ri.deleteObserver(this);
    rj.deleteObserver(this);
    ai.deleteObserver(this);
    aj.deleteObserver(this);
    Index.remove(pos);
  }
  
  public static PartialEdge At(Vertex vertex, Edge edge) {
    TriPosition T = new TriPosition(vertex.getSerialNumber(), edge.getSerialNumber());
    PartialEdge p = Index.get(T);
    if (p == null) {
      p = new PartialEdge(vertex, edge);
      p.pos = T;
      Index.put(T, p);
    }
    return p;
  }
  
  public static double valueAt(Vertex vertex, Edge edge) {
    return At(vertex, edge).getValue();
  }
  
  public PartialEdge.Partial partialAt(Edge f) {
    TriPosition T = new TriPosition(f.getSerialNumber());
    Partial q = PartialIndex.get(T);
    if(q == null) {
      q = new Partial(f);
      q.pos = T;
      PartialIndex.put(T, q);
    }
    return q;
  }
  
  public class Partial extends Geoquant {
    private boolean equal;
    private Eta eij;
    
    private Partial(Edge f) {
      super(f);
      equal = e == f;
      if(equal) {
        ri.addObserver(this);
        rj.addObserver(this);
        ai.addObserver(this);
        aj.addObserver(this);
        eij = Eta.At(e);
        eij.addObserver(this);
      }
    }
    
    protected void recalculate() {
      value = 0;
      if(equal) {
        double radi = ri.getValue();
        double radj = rj.getValue();
        double alphi = ai.getValue();
        double alphj = aj.getValue();
        double eta = eij.getValue();

        value = radi * Math.pow(radj, 2) * (eta * radi + alphj * radj) 
           / Math.pow(alphi * Math.pow(radi, 2) + alphj * Math.pow(radj, 2) + 2*eta*radi*radj, 1.5);
      }
    }

    public void remove() {
      deleteDependents();
      if(equal) {
        ri.deleteObserver(this);
        rj.deleteObserver(this);
        ai.deleteObserver(this);
        aj.deleteObserver(this);
        eij.deleteObserver(this);
      }
      PartialIndex.remove(pos);
    }
    
    public String toString() {
      return "PartialEdge@[" + v + ", " + e + "]" + "w.r.t" + location + "=" + getValue();
    }
  }

  public PartialEdge.SecondPartial secondPartialAt(Edge f, Edge g) {
    TriPosition T = new TriPosition(f.getSerialNumber(), g.getSerialNumber());
    SecondPartial q = SecondPartialIndex.get(T);
    if(q == null) {
      q = new SecondPartial(f, g);
      q.pos = T;
      SecondPartialIndex.put(T, q);
    }
    return q;
  }
  
  public class SecondPartial extends Geoquant {
    private boolean equal;
    private Eta eij;
    
    private SecondPartial(Edge f, Edge g) {
      super(f, g);
      equal = e == f;
      equal = equal && e == g;
      if(equal) {
        ri.addObserver(this);
        rj.addObserver(this);
        ai.addObserver(this);
        aj.addObserver(this);
        eij = Eta.At(e);
        eij.addObserver(this);
      }
    }

    protected void recalculate() {
      value = 0;
      if(equal) {
        double r1 = ri.getValue();
        double r2 = rj.getValue();
        double alpha1 = ai.getValue();
        double alpha2 = aj.getValue();
        double Eta12 = eij.getValue();

        value = (Math.pow(r1,2)*Math.pow(r2,2)*(alpha1*Math.pow(r1,2) - r2*(Eta12*r1 + 2*alpha2*r2)))/
          Math.pow(alpha1*Math.pow(r1,2) + r2*(2*Eta12*r1 + alpha2*r2),2.5);
      }
    }

    public void remove() {
      deleteDependents();
      if(equal) {
        ri.deleteObserver(this);
        rj.deleteObserver(this);
        ai.deleteObserver(this);
        aj.deleteObserver(this);
        eij.deleteObserver(this);
      }
      SecondPartialIndex.remove(pos);
    }
    
    public String toString() {
      return "PartialEdge@[" + v + ", " + e + "]" + "w.r.t" + location + "=" + getValue();
    }
    
  }
}
